File: video_42_PrintingLongestIncreasingSubsequence.cpp
// #include<iostream>
// #include<bits/stdc++.h>
// using namespace std;

/********************************************** LENGTH OF LIS *********************************************** */
// int longestIncreasingSubsequence(int arr[], int n){
//     // Write Your Code here
//     int maxi = 1;
//     vector<int> dp(n, 1);

//     for(int ind = 0; ind<n; ind++){
//         for(int prev_ind = 0; prev_ind<ind; prev_ind++){
//             if(arr[prev_ind]<arr[ind]){
//                 dp[ind] = max(dp[ind], 1+dp[prev_ind]);
//             }
//         }maxi = max(maxi, dp[ind]);
//     }return maxi;
// }

/********************************************** PRINTING THE LIS ******************************************** */
// int longestIncreasingSubsequence(int arr[], int n){
//     // Write Your Code here
//     int maxi = 1, lastInd = 0;
//     vector<int> dp(n, 1), mp(n, 0);

//     for(int ind = 0; ind<n; ind++){
//         mp[ind] = ind;
//         for(int prev_ind = 0; prev_ind<ind; prev_ind++){
//             if(arr[prev_ind]<arr[ind]){
//                 if(1+dp[prev_ind]>dp[ind]){
//                     dp[ind] = 1+dp[prev_ind];
//                     mp[ind] = prev_ind;
//             }
//         }
//         if(dp[ind]>maxi){
//             maxi = dp[ind];
//             lastInd = ind;
//         }
//     }
//     while(mp[lastInd] != lastInd)
//     return maxi;
// }

// int main(){
//     int n, cin>>n;
//     int *arr = new int[n];
//     for(int i = 0; i<n; i++) cin>>arr[i];
//     cout<<longestIncreasingSubsequence(arr, n);
//     return 0;
// }

File: video_42_Printing_Longest_Increasing_Subsequence.java
import java.util.*;
class video_42_Printing_Longest_Increasing_Subsequence {
    public static ArrayList<Integer> longestIncreasingSubsequence(int n, int arr[]) {
        // Code here
        int dp[] = new int[n];
        Arrays.fill(dp, 1);
        int maxi = 1;
        for(int ind = 0; ind<n-1; ind++){
            for(int prevInd = 0; prevInd<ind; prevInd++){
                if(arr[ind]>arr[prevInd]){
                    dp[ind] = Math.max(1+dp[prevInd], dp[ind]);
                }
            }
            maxi = Math.max(maxi, dp[ind]);
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int arr[] = new int[n];

        for(int i = 0; i<n; i++)
            arr[i] = sc.nextInt();
        ArrayList<Integer> ans = longestIncreasingSubsequence(n, arr);
        for(int it: ans)
            System.out.print(it+" ");
    }
}